### PR TITLE
Mark store as corrupted if its translog is bad

### LINKED ISSUE #16662: Translog recovery failure eating CPU
Found one of the nodes in my idle (other than marvel), local **2.2.0** cluster eating a chunk of CPU today.

image

```
% curl -s localhost:9200/_cat/nodes\?h=ip,master,name
127.0.0.1 * Graviton     
127.0.0.1 m Blood Spider 
% curl -s localhost:9200/_cat/shards | fgrep 02.13
.marvel-es-2016.02.13                     0 p INITIALIZING                127.0.0.1 Graviton     
.marvel-es-2016.02.13                     0 r UNASSIGNED
%
```

A bunch of checkpoint logs happen:

```
[2016-02-14 00:29:10,769][DEBUG][index.translog           ] [Graviton] [.marvel-es-2016.02.13][0] recovered local translog from checkpoint Checkpoint{offset=2556515, numOps=4273, translogFileGe
neration= 3691}
```

Both nodes load their data dir from the same `nodes/` path, and this volume is **87% full**.  So allocation of this shard keeps occurring:

```
[2016-02-14 00:29:28,147][INFO ][cluster.routing.allocation.decider] [Graviton] low disk watermark [85%] exceeded on [DG6Jls2UTrSSsrG2huoYMQ][Graviton][/home/aar/Downloads/elasticsearch-2.2.0/d
ata/elasticsearch/nodes/1] free: 66.2gb[14.8%], replicas will not be assigned to this node
[2016-02-14 00:29:28,147][INFO ][cluster.routing.allocation.decider] [Graviton] low disk watermark [85%] exceeded on [w8hx8OmwRg6ETfhbAA6XYg][Blood Spider][/home/aar/Downloads/elasticsearch-2.2
.0/data/elasticsearch/nodes/0] free: 66.2gb[14.8%], replicas will not be assigned to this node
```

Then the allocation is attempted:

```
[2016-02-14 00:29:29,120][DEBUG][index.translog           ] [Graviton] [.marvel-es-2016.02.13][0] translog closed
[2016-02-14 00:29:29,120][DEBUG][index                    ] [Graviton] [.marvel-es-2016.02.13] [0] closing... (reason: [failed recovery])
[2016-02-14 00:29:29,120][DEBUG][index.shard              ] [Graviton] [.marvel-es-2016.02.13][0] state: [RECOVERING]->[CLOSED], reason [failed recovery]
[2016-02-14 00:29:29,120][DEBUG][index.shard              ] [Graviton] [.marvel-es-2016.02.13][0] operations counter reached 0, will not accept any further writes
[2016-02-14 00:29:29,120][DEBUG][index.store              ] [Graviton] [.marvel-es-2016.02.13][0] store reference count on close: 0
[2016-02-14 00:29:29,120][DEBUG][index                    ] [Graviton] [.marvel-es-2016.02.13] [0] closed (reason: [failed recovery])
[2016-02-14 00:29:29,120][WARN ][indices.cluster          ] [Graviton] [[.marvel-es-2016.02.13][0]] marking and sending shard failed due to [failed recovery]
[.marvel-es-2016.02.13][[.marvel-es-2016.02.13][0]] IndexShardRecoveryException[failed to recovery from gateway]; nested: EngineCreationFailureException[failed to recover from translog]; nested: EngineException[failed to recover from translog]; nested: ElasticsearchException[unexpected exception reading from translog snapshot of /home/aar/Downloads/elasticsearch-2.2.0/data/elasticsearch/nodes/1/indices/.marvel-es-2016.02.13/0/translog/translog-6.tlog]; nested: EOFException[read past EOF. pos [2556515] length: [4] end: [2556515]];
        at org.elasticsearch.index.shard.StoreRecoveryService.recoverFromStore(StoreRecoveryService.java:254)
        at org.elasticsearch.index.shard.StoreRecoveryService.access$100(StoreRecoveryService.java:56)
        at org.elasticsearch.index.shard.StoreRecoveryService$1.run(StoreRecoveryService.java:129)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: [.marvel-es-2016.02.13][[.marvel-es-2016.02.13][0]] EngineCreationFailureException[failed to recover from translog]; nested: EngineException[failed to recover from translog]; nested: ElasticsearchException[unexpected exception reading from translog snapshot of /home/aar/Downloads/elasticsearch-2.2.0/data/elasticsearch/nodes/1/indices/.marvel-es-2016.02.13/0/translog/translog-6.tlog]; nested: EOFException[read past EOF. pos [2556515] length: [4] end: [2556515]];
        at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:178)
        at org.elasticsearch.index.engine.InternalEngineFactory.newReadWriteEngine(InternalEngineFactory.java:25)
        at org.elasticsearch.index.shard.IndexShard.newEngine(IndexShard.java:1450)
        at org.elasticsearch.index.shard.IndexShard.createNewEngine(IndexShard.java:1434)
        at org.elasticsearch.index.shard.IndexShard.internalPerformTranslogRecovery(IndexShard.java:925)
        at org.elasticsearch.index.shard.IndexShard.performTranslogRecovery(IndexShard.java:897)
        at org.elasticsearch.index.shard.StoreRecoveryService.recoverFromStore(StoreRecoveryService.java:245)
        ... 5 more
Caused by: [.marvel-es-2016.02.13][[.marvel-es-2016.02.13][0]] EngineException[failed to recover from translog]; nested: ElasticsearchException[unexpected exception reading from translog snapshot of /home/aar/Downloads/elasticsearch-2.2.0/data/elasticsearch/nodes/1/indices/.marvel-es-2016.02.13/0/translog/translog-6.tlog]; nested: EOFException[read past EOF. pos [2556515] length: [4] end: [2556515]];
        at org.elasticsearch.index.engine.InternalEngine.recoverFromTranslog(InternalEngine.java:254)
        at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:175)
        ... 11 more
Caused by: ElasticsearchException[unexpected exception reading from translog snapshot of /home/aar/Downloads/elasticsearch-2.2.0/data/elasticsearch/nodes/1/indices/.marvel-es-2016.02.13/0/translog/translog-6.tlog]; nested: EOFException[read past EOF. pos [2556515] length: [4] end: [2556515]];
        at org.elasticsearch.index.translog.TranslogReader.readSize(TranslogReader.java:102)
        at org.elasticsearch.index.translog.TranslogReader.access$000(TranslogReader.java:46)
        at org.elasticsearch.index.translog.TranslogReader$ReaderSnapshot.readOperation(TranslogReader.java:297)
        at org.elasticsearch.index.translog.TranslogReader$ReaderSnapshot.next(TranslogReader.java:290)
        at org.elasticsearch.index.translog.MultiSnapshot.next(MultiSnapshot.java:70)
        at org.elasticsearch.index.engine.InternalEngine.recoverFromTranslog(InternalEngine.java:240)
        ... 12 more
Caused by: java.io.EOFException: read past EOF. pos [2556515] length: [4] end: [2556515]
        at org.elasticsearch.common.io.Channels.readFromFileChannelWithEofException(Channels.java:102)
        at org.elasticsearch.index.translog.ImmutableTranslogReader.readBytes(ImmutableTranslogReader.java:84)
        at org.elasticsearch.index.translog.TranslogReader.readSize(TranslogReader.java:91)
        ... 17 more
[2016-02-14 00:29:29,121][WARN ][cluster.action.shard     ] [Graviton] [.marvel-es-2016.02.13][0] received shard failed for [.marvel-es-2016.02.13][0], node[DG6Jls2UTrSSsrG2huoYMQ], [P], v[7374], s[INITIALIZING], a[id=0EodiK9gS8KIGvHSHNi1ew], unassigned_info[[reason=ALLOCATION_FAILED], at[2016-02-14T06:29:10.518Z], details[failed recovery, failure IndexShardRecoveryException[failed to recovery from gateway]; nested: EngineCreationFailureException[failed to recover from translog]; nested: EngineException[failed to recover from translog]; nested: ElasticsearchException[unexpected exception reading from translog snapshot of /home/aar/Downloads/elasticsearch-2.2.0/data/elasticsearch/nodes/0/indices/.marvel-es-2016.02.13/0/translog/translog-6.tlog]; nested: EOFException[read past EOF. pos [2343547] length: [4] end: [2343547]]; ]], indexUUID [Wl7rh0TuT1ms79hg-H8hzA], message [failed recovery], failure [IndexShardRecoveryException[failed to recovery from gateway]; nested: EngineCreationFailureException[failed to recover from translog]; nested: EngineException[failed to recover from translog]; nested: ElasticsearchException[unexpected exception reading from translog snapshot of /home/aar/Downloads/elasticsearch-2.2.0/data/elasticsearch/nodes/1/indices/.marvel-es-2016.02.13/0/translog/translog-6.tlog]; nested: EOFException[read past EOF. pos [2556515] length: [4] end: [2556515]]; ]
```

This process repeats ad infinitum.  `DEBUG` log file is each over 1g the past couple of days.

`hot_threads` has this repeated section:

```
   98.7% (493.3ms out of 500ms) cpu usage by thread 'elasticsearch[Graviton][generic][T#1]'
     10/10 snapshots sharing following 24 elements
       sun.nio.fs.UnixNativeDispatcher.readdir(Native Method)
       sun.nio.fs.UnixDirectoryStream$UnixDirectoryIterator.readNextEntry(UnixDirectoryStream.java:168)
       sun.nio.fs.UnixDirectoryStream$UnixDirectoryIterator.hasNext(UnixDirectoryStream.java:201)
       org.elasticsearch.index.translog.Translog$OnCloseRunnable.handle(Translog.java:726)
       org.elasticsearch.index.translog.Translog$OnCloseRunnable.handle(Translog.java:713)
       org.elasticsearch.index.translog.ChannelReference.closeInternal(ChannelReference.java:67)
       org.elasticsearch.common.util.concurrent.AbstractRefCounted.decRef(AbstractRefCounted.java:64)
       org.elasticsearch.index.translog.TranslogReader.close(TranslogReader.java:143)
       org.apache.lucene.util.IOUtils.close(IOUtils.java:97)
       org.elasticsearch.index.translog.Translog.close(Translog.java:425)
       org.apache.lucene.util.IOUtils.closeWhileHandlingException(IOUtils.java:129)
       org.apache.lucene.util.IOUtils.closeWhileHandlingException(IOUtils.java:118)
       org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:183)
       org.elasticsearch.index.engine.InternalEngineFactory.newReadWriteEngine(InternalEngineFactory.java:25)
       org.elasticsearch.index.shard.IndexShard.newEngine(IndexShard.java:1450)
       org.elasticsearch.index.shard.IndexShard.createNewEngine(IndexShard.java:1434)
       org.elasticsearch.index.shard.IndexShard.internalPerformTranslogRecovery(IndexShard.java:925)
       org.elasticsearch.index.shard.IndexShard.performTranslogRecovery(IndexShard.java:897)
       org.elasticsearch.index.shard.StoreRecoveryService.recoverFromStore(StoreRecoveryService.java:245)
       org.elasticsearch.index.shard.StoreRecoveryService.access$100(StoreRecoveryService.java:56)
       org.elasticsearch.index.shard.StoreRecoveryService$1.run(StoreRecoveryService.java:129)
       java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
       java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
       java.lang.Thread.run(Thread.java:745)
```

### PR DESCRIPTION
**Pull request by dnhatn**
Today if Lucene index is not good, we mark its store as corrupted by storing a corruption marker. With the presence of the corruption marker, a corrupted copy won't be used as a primary and requires a full file-based recovery to be used a replica. This restriction should also apply to a copy with corrupted translog. If a copy with corrupted translog is mistakenly selected as a primary, it may not be able to provide all translog operations properly for replicas in the future.

This commit marks the store of a shard copy as corrupted if its translog is not good.

Relates #28435
Closes #16662

### CODE REVIEW

**Comment by dnhatn** [comment]
This issue was caught by `#testReplicaCorruption 
` in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-windows-compatibility/1179/console (Log: [testReplicaCorruption.txt](https://github.com/elastic/elasticsearch/files/1676317/testReplicaCorruption.txt)).

We intentionally corrupted every segment files in `#testReplicaCorruption`. This caused file-based recovery occurring as replicas could not to read commit snapshot. However, there was one replica `[test][7]` failed to clean up its store after index files were copied. The corruption marker `corrupted_WzIGIIeIQvqVgKvEIVtRwA` was deleted but still in the listing and inaccessible(on Windows).

```
1> Caused by: java.nio.file.AccessDeniedException: C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\server\build\testrun\integTest\J1\temp\org.elasticsearch.index.store.CorruptedFileIT_883F8FC2C6E448B5-001\tempDir-002\data\nodes\0\indices\0tscMvZ-Q-21DRHviJsJxw\7\index\corrupted_WzIGIIeIQvqVgKvEIVtRwA
  1> 	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:83) ~[?:?]
  1> 	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97) ~[?:?]
  1> 	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102) ~[?:?]
  1> 	at sun.nio.fs.WindowsFileSystemProvider.newByteChannel(WindowsFileSystemProvider.java:230) ~[?:?]
  1> 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newByteChannel(FilterFileSystemProvider.java:212) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newByteChannel(FilterFileSystemProvider.java:212) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newByteChannel(FilterFileSystemProvider.java:212) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.HandleTrackingFS.newByteChannel(HandleTrackingFS.java:240) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newByteChannel(FilterFileSystemProvider.java:212) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.mockfile.HandleTrackingFS.newByteChannel(HandleTrackingFS.java:240) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at java.nio.file.Files.newByteChannel(Files.java:361) ~[?:1.8.0_144]
  1> 	at java.nio.file.Files.newByteChannel(Files.java:407) ~[?:1.8.0_144]
  1> 	at org.apache.lucene.store.SimpleFSDirectory.openInput(SimpleFSDirectory.java:77) ~[lucene-core-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:43]
  1> 	at org.apache.lucene.util.LuceneTestCase.slowFileExists(LuceneTestCase.java:2709) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:44]
  1> 	at org.apache.lucene.store.MockDirectoryWrapper.openInput(MockDirectoryWrapper.java:749) ~[lucene-test-framework-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:43]
  1> 	at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:99) ~[lucene-core-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:43]
  1> 	at org.apache.lucene.store.Directory.openChecksumInput(Directory.java:119) ~[lucene-core-7.2.1.jar:7.2.1 b2b6438b37073bee1fca40374e85bf91aa457c0b - ubuntu - 2018-01-10 00:48:43]
  1> 	at org.elasticsearch.index.store.Store.failIfCorrupted(Store.java:615) ~[main/:?]
  1> 	at org.elasticsearch.index.store.Store.failIfCorrupted(Store.java:607) ~[main/:?]
  1> 	at org.elasticsearch.index.store.Store.getMetadata(Store.java:281) ~[main/:?]
  1> 	at org.elasticsearch.index.store.Store.getMetadata(Store.java:253) ~[main/:?]
  1> 	at org.elasticsearch.index.store.Store.cleanupAndVerify(Store.java:683) ~[main/:?]
  1> 	at org.elasticsearch.indices.recovery.RecoveryTarget.cleanFiles(RecoveryTarget.java:444) [main/:?]
  1> 	... 10 more
```

As the target failed to clean up, it retried another recovery. At that moment the corrupted marker was removed and the copied commit is safe, hence the replica starts sequence-based recovery. Unfortunately, translog and index commit on the target were mismatch because the Lucene commit was copied from the primary. This replica `[test][7]` could not recover from primary.

```
  1> Caused by: org.elasticsearch.index.translog.TranslogCorruptedException: expected shard UUID [32 5f 7a 61 42 43 56 4d 53 4f 47 4a 6a 35 30 4f 6c 69 65 43 44 51] but got: [54 62 51 6c 30 6a 49 4f 53 50 53 68 5a 76 58 2d 2d 44 46 70 58 41] this translog file belongs to a different translog. path:C:\Users\jenkins\workspace\elastic+elasticsearch+master+multijob-windows-compatibility\server\build\testrun\integTest\J1\temp\org.elasticsearch.index.store.CorruptedFileIT_883F8FC2C6E448B5-001\tempDir-002\data\nodes\0\indices\0tscMvZ-Q-21DRHviJsJxw\7\translog\translog-2.tlog
  1> 	at org.elasticsearch.index.translog.TranslogReader.open(TranslogReader.java:131) ~[main/:?]
  1> 	at org.elasticsearch.index.translog.Translog.openReader(Translog.java:292) ~[main/:?]
  1> 	at org.elasticsearch.index.translog.Translog.recoverFromFiles(Translog.java:238) ~[main/:?]
  1> 	at org.elasticsearch.index.translog.Translog.<init>(Translog.java:176) ~[main/:?]
  1> 	at org.elasticsearch.index.engine.InternalEngine.openTranslog(InternalEngine.java:491) ~[main/:?]
  1> 	at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:180) ~[main/:?]
  1> 	at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:151) ~[main/:?]
  1> 	at org.elasticsearch.index.engine.InternalEngineFactory.newReadWriteEngine(InternalEngineFactory.java:25) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.newEngine(IndexShard.java:2149) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.createNewEngine(IndexShard.java:2131) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.innerOpenEngineAndTranslog(IndexShard.java:1361) ~[main/:?]
  1> 	at org.elasticsearch.index.shard.IndexShard.openIndexAndSkipTranslogRecovery(IndexShard.java:1327) ~[main/:?]
  1> 	at org.elasticsearch.indices.recovery.RecoveryTarget.prepareForTranslogOperations(RecoveryTarget.java:371) ~[main/:?]
  1> 	at org.elasticsearch.indices.recovery.PeerRecoveryTargetService$PrepareForTranslogOperationsRequestHandler.messageReceived(PeerRecoveryTargetService.java:396) ~[main/:?]
  1> 	at org.elasticsearch.indices.recovery.PeerRecoveryTargetService$PrepareForTranslogOperationsRequestHandler.messageReceived(PeerRecoveryTargetService.java:390) ~[main/:?]
```

**Comment by bleskes** [comment]
Thanks @dnhatn . I'm not 100% sure about this approach - it might be the right one but there are problems with it  - like what does a store corruption mean? does it mean that the lucene files are not good or does it mean that the shard is not good? If the lucene files are not good we need to do a full file based copy and not rely on the checksums for comparison. If  the lucene files are good and the translog is bad, we can reuse files for file based copy but not do a seq# recovery. 

My initial response is to say that we should not be overly heroic and have one marker that says "shard storage good or bad". If it's good it can be used for any purpose, if it's bad - it can't. This is something we should discuss with a bigger group first.

Finally - we can have an alternative solution as well - just putting it out there (it has complexity as well to implement) - we can validate the that the translog we read the global checkpoint actually belongs to the lucene index that we search for safe commits in.

**Comment by dnhatn** [comment]
@bleskes, I've updated the PR to mark the store as corrupted whenever we detect that the translog is bad. Could you please take a look? Thank you!

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -1345,15 +1346,20 @@ private void innerOpenEngineAndTranslog(final EngineConfig.OpenMode openMode, fi
...
+                // we have to set it before we open an engine and recover from the translog because
+                // acquiring a snapshot from the translog causes a sync which causes the global checkpoint to be pulled in,
+                // and an engine can be forced to close in ctor which also causes the global checkpoint to be pulled in.
+                final String translogUUID = store.readLastCommittedSegmentsInfo().getUserData().get(Translog.TRANSLOG_UUID_KEY);
+                final long globalCheckpoint = Translog.readGlobalCheckpoint(translogConfig.getTranslogPath(), translogUUID);
+                replicationTracker.updateGlobalCheckpointOnReplica(globalCheckpoint, "read from translog checkpoint");
+            }
+            createNewEngine(config);
+        } catch (TranslogCorruptedException ex) {
+            store.markStoreCorrupted(new IOException("failed to open engine; translog corrupted", ex));
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I wonder if we shoudl move this to the engine constructor. We already have catch blocks there.

File: server/src/main/java/org/elasticsearch/index/translog/TruncateTranslogCommand.java [hunk 2]
```diff
@@ -192,6 +193,8 @@ protected void execute(Terminal terminal, OptionSet options, Environment env) th
                 writer.setLiveCommitData(newCommitData.entrySet());
                 writer.commit();
             }
+            Store.removeCorruptionMarker(dir);
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/translog/TruncateTranslogCommand.java]: This should only remove translog corruptions.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 3]
```diff
@@ -469,13 +470,17 @@ long phase2(final long startingSeqNo, long requiredSeqNoRangeStart, long endingS
...
-        response.phase2Operations = result.totalOperations;
-        return result.targetLocalCheckpoint;
+        try {
+            final SendSnapshotResult result = sendSnapshot(startingSeqNo, requiredSeqNoRangeStart, endingSeqNo, snapshot);
+            stopWatch.stop();
+            logger.trace("recovery [phase2]: took [{}]", stopWatch.totalTime());
+            response.phase2Time = stopWatch.totalTime().millis();
+            response.phase2Operations = result.totalOperations;
+            return result.targetLocalCheckpoint;
+        } catch (TranslogCorruptedException ex) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I think this should somehow be folded into the snapshot reading. This means the engine will be failed regardless on who uses the snapshot.

**Review by bleskes — COMMENTED** [review]
I left initial comments

**Comment by dnhatn** [comment]
As discussed with @bleskes on another channel, we need to isolate the translog from an engine before proceeding this.

**Comment by dnhatn** [comment]
@bleskes I am closing this for now and reopen when it's ready again. Thanks for looking at this.

**Pull request closed without merging**